### PR TITLE
added the restricted keywords

### PR DESCRIPTION
**Pull request by borissmidt**
Dear Dennis, 
I have added a restricted and restricted_const keyword. however i have not discovered any increase in performance when using these.

i also added an extra cmakelist which only adds the vexcl files so they are visible if you use the cmakelist as a project file. 
 
I edited some lines of the reduction kernel. so it doesn't copy from smem to sdata to smem again since this was a useless copy and it saved a single instruction hurray :) while not breaking the code. 

Sincerely, Boris Smidt.

### CODE REVIEW

File: vexcl/reductor.hpp [hunk 1]
```diff
@@ -326,28 +326,27 @@ Reductor<real,RDC>::operator()(const Expr &expr) const {
                             queue[d], source.str(), "vexcl_reductor_kernel"));
             } else {
                 source.smem_declaration<real>();
-                source.new_line() << type_name< shared_ptr<real> >() << " sdata = smem;";
```
Comment (ddemidov) [inline vexcl/reductor.hpp]: This does break the CUDA backend. In CUDA, as opposed to OpenCL, the shared memory is not passed as a kernel parameter, it is just declared as an array with `__shared__` qualifier. Compare implementation of `source.smem_declaration()` (used in the line above) for [OpenCL](https://github.com/ddemidov/vexcl/blob/7ac0b2516bc3317abf9893a5afb0828a05a93c6d/vexcl/backend/opencl/source.hpp#L197-L201) and for [CUDA](https://github.com/ddemidov/vexcl/blob/7ac0b2516bc3317abf9893a5afb0828a05a93c6d/vexcl/backend/cuda/source.hpp#L176-L180) backends.

File: vexcl/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
+FILE(GLOB_RECURSE MyCSources *.hpp )
+add_library(vex ${MyCSources})
+
+SET_TARGET_PROPERTIES(vex PROPERTIES LINKER_LANGUAGE CXX)
```
Comment (ddemidov) [inline vexcl/CMakeLists.txt]: I don't understand this change. There is no library to build here, vexcl is header-only. One only need to add vexcl root directory to include paths with a compiler switch.

File: README.md [hunk 3]
```diff
@@ -376,6 +377,12 @@ Z = sqrt(tag<1>(X) * tag<1>(X) + tag<2>(Y) * tag<2>(Y));
 Here, the generated kernel will have one parameter for each of the vectors `X`
 and `Y`.
 
+### <a name="optimisation-keyword"></a>Optimisation Keywords
```
Comment (ddemidov) [inline README.md]: 'Optimization keywords' does not sound right. I think that 'pointer qualifiers' is a better name, but I am not about that either.

File: vexcl/optimisation.hpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
+#ifndef OPTIMISATIONS_HPP
```
Comment (ddemidov) [inline vexcl/optimisation.hpp]: I don't think that 'optimization' is a good name for this functionality. What about 'qualifiers' or even 'pointer_qualifiers'? Anyway, to make it consistent with the rest of the code, the include guard should contain the full path to the file from vexcl root folder. (as in `VEXCL_OPTIMIZATIONS_HPP`).

**Comment by borissmidt** [comment]
ok i will change those things.

File: vexcl/optimisation.hpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+#define OPTIMISATIONS_HPP
+
+
+#include <vexcl/operations.hpp>
+
+namespace vex {
+//_________________________________________________ restrict keyword
```
Comment (ddemidov) [inline vexcl/optimisation.hpp]: Can you please enclose the supporting code (that is not indended for direct user consumption) between
```
/// \cond INTERNAL
```
and 
```
/// \endcond
```
comments? See [tagged_terminal.hpp](https://github.com/ddemidov/vexcl/blob/7ac0b2516bc3317abf9893a5afb0828a05a93c6d/vexcl/tagged_terminal.hpp#L44) for an example. This would help Doxygen to  skip unneeded code during documentation generation.

File: vexcl/optimisation.hpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+// This struct will hold a reference to a vector that needs to be declared with
+// 'restrict' qualifier.
+template <class T>
+struct restricted_vector : restricted_vector_expression {
+typedef T value_type; // has to be here for vex::deduce() to work with restricted vectors.
+const vector<T> &v;
+    restricted_vector(const vector<T> &v) : v(v) {}
+};
+
+// A convenience function to be used in user code.
```
Comment (ddemidov) [inline vexcl/optimisation.hpp]: Can you make this a Doxygen comment? See [here](https://github.com/ddemidov/vexcl/blob/7ac0b2516bc3317abf9893a5afb0828a05a93c6d/vexcl/tagged_terminal.hpp#L253-L264) for an example.

File: vexcl/optimisation.hpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+// proto::value() function.
+template <>
+struct proto_terminal_is_value< restricted_vector_terminal > : std::true_type {};
+
+// Declare kernel parameter for the restricted vector.
+// This is the only specification that is different from vector<T>.
+template <typename T>
+struct kernel_param_declaration< restricted_vector<T> > {
+static void get(backend::source_generator &src,
+const restricted_vector<T>&,
```
Comment (ddemidov) [inline vexcl/optimisation.hpp]: Can you add an indentation here and through the rest of the file (consistent with the rest of the VexCL code)?

**Comment by ddemidov** [comment]
Thank you! I've left some notes in the code.

File: vexcl/optimisation.hpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
+#ifndef OPTIMISATIONS_HPP
+#define OPTIMISATIONS_HPP
+
+
```
Comment (ddemidov) [inline vexcl/optimisation.hpp]: Can you add a copyright notice here, similar to the rest of the header files? (you should put your name there).

File: CMakeLists.txt [hunk 9]
```diff
@@ -131,9 +131,9 @@ endif ()
 #----------------------------------------------------------------------------
 enable_testing()
 add_subdirectory(tests)
-
 add_subdirectory(examples)
 add_subdirectory(doc)
 add_subdirectory(cmake)
 
+
```
Comment (ddemidov) [inline CMakeLists.txt]: Can you revert this as well?

File: README.md [hunk 10]
```diff
@@ -38,6 +38,7 @@ Other talks may be found at
     * [Element indices](#element-indices)
     * [User-defined functions](#user-defined-functions)
     * [Tagged terminals](#tagged-terminals)
+	* [Pointer qualifiers](#pointer-qualifiers)
```
Comment (ddemidov) [inline README.md]: I think it should be on the same level with 'Tagged terminals'

File: README.md [hunk 11]
```diff
@@ -376,6 +377,12 @@ Z = sqrt(tag<1>(X) * tag<1>(X) + tag<2>(Y) * tag<2>(Y));
 Here, the generated kernel will have one parameter for each of the vectors `X`
 and `Y`.
 
+### <a name="pointer-qualifiers"></a>pointer qualifiers
```
Comment (ddemidov) [inline README.md]: Should be '**P**ointer qualifiers'

**Comment by borissmidt** [comment]
i seems to have a problem to think of an example since its not yet possible to use these keywords qualifiers with a raw pointer. 
but for example this loop body of a kernel would be optimized when you tell the compiler you are not aliasing. of course it is possible to add them all at once :). 
~~~{.c}
prm_2[idx] +=  prm_3[idx] ;
prm_2[idx] +=  prm_3[idx-1] ;
prm_2[idx] +=  prm_3[idx+1] ;
~~~

File: README.md (hunk 10 above)
Comment (borissmidt) [inline README.md]: i see i used a tab.

File: README.md [hunk 12]
```diff
@@ -376,6 +377,12 @@ Z = sqrt(tag<1>(X) * tag<1>(X) + tag<2>(Y) * tag<2>(Y));
...
+VexCL has implmented some of the pointer qualifiers for kernel input variables.
+The programmer can now tag non ailising pointer with the 'restricted', this adds the 'restrict' keyword to the pointer,
+or and constant arrays 'restricted_const' which results is 'const type * restrict'.
+
```
Comment (ddemidov) [inline README.md]: There are couple of typos here. (implmented, ailising).
I would rephrase it anyway, may be to
```
VexCL allows the user to qualify the pointers passed to OpenCL/CUDA kernels with 'restrict' keyword, which limits the effects of pointer aliasing, aiding optimizations. This is achieved by wrapping a vector in an expression with either `restricted()` or `const_restricted()` function.
```

Comment (ddemidov) [inline README.md]: Also, an example would be helpful here.

**Comment by ddemidov** [comment]
This should work with vector instances and not with raw_pointer.

A possible example is
```cpp
x = 2 * restricted(y);
```
where both `x` and `y` are instances of a `vex::vector<T>`.

**Comment by ddemidov** [comment]
Since we are talking about examples, could you please add a test case for this functionality?

**Comment by ddemidov** [comment]
On a second thought, this could indeed be useful with 'raw_pointer()', and there is currently no way to combine `raw_pointer()` and `restricted()` functions. I think we should forget about `raw_pointer()` for now.

**Comment by borissmidt** [comment]
i'm really fighting with git again :(. 

and this example results in the same code  (except for the restricted_const which might be cached)
x = 2 * restricted(y);

it has to be a combination of 2 lines to result in an ailising problem. 
like 
x+=2 * y
x+=2 * y
if this ailises then x can be y and this y 
but with it could be 2 * y +  (x+2*y). 
but when ailising is not allowed : x + 2*y +2*y

File: README.md [hunk 13]
```diff
@@ -376,6 +377,12 @@ Z = sqrt(tag<1>(X) * tag<1>(X) + tag<2>(Y) * tag<2>(Y));
 Here, the generated kernel will have one parameter for each of the vectors `X`
 and `Y`.
 
+### <a name="pointer-qualifiers"></a>Pointer qualifiers
+VexCL has implmented some of the pointer qualifiers for kernel input variables.
+The programmer can now tag non ailising pointer with the 'restricted', this adds the 'restrict' keyword to the pointer,
+or and constant arrays 'restricted_const' which results is 'const type * restrict'.
+
```
Comment (ddemidov) [inline README.md]: Excuse me for repeating myself, I already made this comment before,  I think it got lost in outdated diffs:

There are couple of typos here. (implmented, ailising). And I would rephrase the paragraph it anyway, may be to

VexCL allows the user to qualify the pointers passed to OpenCL/CUDA kernels with 'restrict' keyword, which limits the effects of pointer aliasing, aiding optimizations. This is achieved by wrapping a vector in an expression with either restricted() or const_restricted() function.

**Comment by ddemidov** [comment]
Some thoughts about `raw_pointer` problem. Currently there are three places where a pointer to an array may appear in OpenCL/CUDA kernel.

* `vex::vector<T>` terminals in expressions. Those are covered by your RP.

* Custom function arguments, as in
```cpp
VEX_FUNCTION(int, dummy, (int, i)(int*, ptr), return ptr[i]; );
x = dummy(element_index(), raw_pointer(y));
```

It seems that `__restrict__` keyword in g++ does change the type of the pointer, as the following example shows:
```cpp
#include <iostream>
#include <type_traits>

int main() {
    std::cout << std::is_same<int*, int*>::value << std::endl;                // 1
    std::cout << std::is_same<int*, int* __restrict__>::value << std::endl;   // 0
}
```

So it should be possible to implement
```cpp
VEX_FUNCTION(int, dummy, (int, i)(int* __restrict__, ptr), return ptr[i]; );
x = dummy(element_index(), raw_pointer(y));
```
The problem here is that `__restrict__` is not standard keyword in C++, and it will work only in g++. Visual studio uses `__restrict` for this purpose.

Another option is to provide a meta function, e.g. `restricted_` (note the trailing underscore to differentiate from `restricted()` that you already implemented):
```cpp
VEX_FUNCTION(int, dummy, (int, i)(restrict_<int*>, ptr), return ptr[i]; );
x = dummy(element_index(), raw_pointer(y));
```
This should work across compilers.

* The `raw_pointer(vex::vector<T>)` expression that is not passed to any function but is used directly. An example is
```cpp
x = *(raw_pointer(y) + 1); // equivalent to x[:] = y[1]
```

I think it would be natural to use the same function you have already here:
```cpp
x = *(restricted(raw_pointer(y)) + 1); // equivalent to x[:] = y[1]
```
I think its possible to implement, but need to think on the details a bit more.

Do you agree with these points? In case you do, would you like to implement them?

**Comment by ddemidov** [comment]
Regarding the example. I understood that it is difficult to construct a meaningful example where restrict keyword would be useful.

There are several ways in vexcl to launch a kernel that does more than one vector assignment at once.  First, there are multivector expressions, e.g.
```cpp
vex::tie(x, y) = std::tie(expr1, expr2);
```
This is equivalent to
```cpp
x = expr1;
y = expr2;
```
Unfortunately, the actual implementation makes it impossible to use for aliasing example. Its implemented as
```cpp
tmp1 = expr1;
tmp2 = expr2;
x = tmp1;
y = tmp2; 
```
So `vex::tie(x, x) = std::tie(y, y)` will always work like `x = y`.

Another option is to use custom function with side effect:
```cpp
VEX_FUNCTION(int, dummy, (int, i)(restricted_<int*>, x)(restricted_<int*>, y),
  x[i] += y[i];
  x[i] += y[i];
  return x[i];
);
z = dummy(element_index(), raw_pointer(x), raw_pointer(y));
```
But it feels both ugly and contrived.

**Comment by borissmidt** [comment]
Maybe a cleaner way would be to implement a 'restricted_pointer()' since it doesn't make sense to be able to change from a normal pointer to a restricted pointer. Actually it seems impossible with current vexCL to do multiple stores in a single kernel .  
Except for custom functions and kernels. If you define a temp it will always be an register (normal) variable and not a global one. 

I just seem to have discovered that openCL declares two kinds of constants a 'global const' an  'constant'. The latter is set in the constant memory cache (32kB in size). While the former doesn't.

OFFTOPIC:
At the moment i am programming a custom kernel to test the performance of mine application. I am going to test the use of the 'native_functions' instead of the normal functions, for example the 'cos' becomes 'native_cos'. vex cl still takes away the annoying boiler plate code and i'm thankful for that.

**Comment by ddemidov** [comment]
I like the idea with `restricted_pointer()`.

Re native functions: You can use any builtin OpenCL/CUDA function not yet provided by VexCL using VEX_BUILTIN_FUNCTION macro in your code. For example:
```cpp
VEX_BUILTIN_FUNCTION( 1, native_cos )

// ...

x = native_cos(y);
```

**Comment by borissmidt** [comment]
Ok thanks, i didn't know. The custom kernel is already written.
But that might be future functionality. Something to put onto your roadmap.

**Comment by ddemidov** [comment]
Do you still think that providing ability to use `restrict` keyword would be of possible use in VexCL?

**Comment by borissmidt** [comment]
It might not with the current setup from VexCL. But it might be useful to optimise some of your standard function. 
But adding the keyword 'constant' will be useful. This can cause the vector to be put into a constant cache if the size is smaller than 64kB . But i do not know if i could make a raw pointer compatible.

**Comment by ddemidov** [comment]
In that case, please 

1. Undo any changes to CMakeLists.txt and vexcl/reductor.hpp, 
2. Remove restrict keyword support from vexcl/pointer_qualifiers.hpp (so just const keyword support is there), and
3. Update the changes to README accordingly.

After that, i think its ready to be merged. I'll add support for raw_pointer() myself after the PR is merged.

**Comment by ddemidov** [comment]
Oh, and please rebase the PR on the latest master and squash all the commits together.
